assets/js/shared/hocs/with-product-data-context.js

```diff
@@ -1,54 +1,60 @@
 /**
  * External dependencies
  */
-import apiFetch from '@wordpress/api-fetch';
 import {
 	ProductDataContextProvider,
 	useProductDataContext,
 } from '@woocommerce/shared-context';
-import { useState, useEffect } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_NAME } from './store';
 
 /**
  * Loads the product from the API and adds to the context provider.
  *
  * @param {Object} props Component props.
  */
 const OriginalComponentWithContext = ( props ) => {
-	const { productId, OriginalComponent } = props;
-	const [ product, setProduct ] = useState( null );
-	const [ isLoading, setIsLoading ] = useState( true );
+	const { productId, OriginalComponent, postId, product } = props;
 
-	useEffect( () => {
-		if ( !! props.product ) {
-			setProduct( props.product );
-			setIsLoading( false );
-		}
-	}, [ props.product ] );
+	const id = props.isDescendentOfQueryLoop ? postId : productId;
 
-	useEffect( () => {
-		if ( productId > 0 ) {
-			setIsLoading( true );
-			apiFetch( {
-				path: `/wc/store/v1/products/${ productId }`,
-			} )
-				.then( ( receivedProduct ) => {
-					setProduct( receivedProduct );
-				} )
-				.catch( async () => {
-					setProduct( null );
-				} )
-				.finally( () => {
-					setIsLoading( false );
-				} );
+	const productFromAPI = useSelect( ( select ) => {
+		if ( id > 0 ) {
+			const prod = select( STORE_NAME ).getProduct( id );
+			const hasFinished = select( STORE_NAME ).hasFinishedResolution(
+				'getProduct',
+				[ id ]
+			);
+
+			return {
+				product: prod || null,
+				isLoading: ! hasFinished,
+			};
 		}
-	}, [ productId ] );
 
-	if ( ! isLoading && ! product ) {
-		return null;
+		return {
+			product: null,
+			isLoading: false,
+		};
+	} );
```

Comment:
Can we use our existing hook here to avoid creating another store? You mentioned we should use `@woocommerce/data`, so I think eventually, we will delete the custom store introduced in this PR. So I think using the existing hook can make it easier to migrate to `@woocommerce/data` later.

I tried to use `useStoreProducts` instead and get the same result: https://github.com/woocommerce/woocommerce-blocks/compare/add/product-image-product-grid...try/use-existing-hook-for-hoc